File: contagent.js
#!/usr/bin/env node
const fs = require("fs");
const os = require("os");
const path = require("path");
const cp = require("child_process");

const USAGE = `Usage: ./contagent.js [options] [--] [command ...]

Options:
  -c, --config CONFIG         Runtime config path (default: .contagent.yaml)
  --<feature>                 Enable volume mounts for an image feature
  --no-<feature>              Disable volume mounts for an image feature
  --show-options              Show config-defined --<feature>/--no-<feature> toggles and exit
  --extra-groups <gid[,gid]>  Append supplementary group GIDs for this run
  -h, --help                  Show this help
`;

const EMBEDDED_CONFIG = "/usr/local/share/contagent/contagent.yaml";
const die = (msg) => { console.error(`ERROR: ${msg}`); process.exit(1); };
const warn = (msg) => console.error(`WARN: ${msg}`);
const id = (flag) => cp.execFileSync("id", [flag], { encoding: "utf8" }).trim();
const hasCmd = (cmd) => cp.spawnSync("sh", ["-lc", `command -v ${cmd}`], { stdio: "ignore" }).status === 0;
const parseJson = (text, msg) => { try { return JSON.parse(text); } catch { die(msg); } };
const isSocket = (p) => { try { return fs.statSync(p).isSocket(); } catch { return false; } };
const resolvePath = (raw, home, base) => raw === "~" || raw.startsWith("~/")
  ? home + raw.slice(1)
  : (path.isAbsolute(raw) ? raw : path.join(base, raw));

function boolValue(value, fallback, feature, field) {
  if (value == null) return fallback;
  if (typeof value !== "boolean") die(`invalid volume entry in feature ${feature}: ${field} must be boolean`);
  return value;
}

function embeddedConfig(image) {
  const proc = cp.spawnSync("docker", ["run", "--rm", "--entrypoint", "cat", image, EMBEDDED_CONFIG], { encoding: "utf8" });
  if (proc.status !== 0) die(`failed to extract default contagent config from ${image}`);
  return proc.stdout;
}

function yqJson(file, image) {
  const attempts = hasCmd("yq")
    ? [["yq", ["-o=json", ".", file], null], ["yq", [".", file], null]]
    : [
        ["docker", ["run", "--rm", "-i", "--entrypoint", "yq", image, "-o=json", ".", "-"], fs.readFileSync(file)],
        ["docker", ["run", "--rm", "-i", "--entrypoint", "yq", image, ".", "-"], fs.readFileSync(file)],
      ];
  if (!hasCmd("yq")) warn(`yq not found; falling back to yq from ${image}`);
  for (const [cmd, args, input] of attempts) {
    const proc = cp.spawnSync(cmd, args, { input, encoding: "utf8" });
    if (proc.status === 0) {
      try { return JSON.parse(proc.stdout); } catch { /* try next form */ }
    }
  }
  die(`failed to parse ${file}`);
}

function configPath(argv) {
  let config = ".contagent.yaml";
  for (let i = 0; i < argv.length; i++) {
    const arg = argv[i];
    if (arg === "--") break;
    if (arg === "-c" || arg === "--config") {
      if (i + 1 >= argv.length) die(`${arg} requires a value`);
      config = argv[++i];
    } else if (arg.startsWith("--config=")) {
      config = arg.split("=", 2)[1];
    } else if (!arg.startsWith("-")) {
      break;
    }
  }
  return config;
}

function loadConfig(file, image) {
  if (!fs.existsSync(file) || fs.statSync(file).size === 0) {
    fs.mkdirSync(path.dirname(file), { recursive: true });
    const data = embeddedConfig(image);
    if (!data) die("embedded default contagent config is empty");
    fs.writeFileSync(file, data);
  }
  const config = yqJson(file, image);
  const tmp = path.join(os.tmpdir(), `contagent-embedded-${process.pid}.yaml`);
  fs.writeFileSync(tmp, embeddedConfig(image));
  const embedded = yqJson(tmp, image);
  fs.rmSync(tmp, { force: true });
  const configId = config["image-hash"] || "";
  const embeddedId = embedded["image-hash"] || "";
  if (embeddedId && configId !== embeddedId) warn(`${file} was not generated from ${image}`);
  return { config, base: path.resolve(path.dirname(file)) };
}

function buildModel(config) {
  const allFeatures = new Set();
  const options = new Map();
  const rows = [];
  const env = [];
  for (const feature of Array.isArray(config?.features) ? config.features : []) {
    const name = String(feature?.name || "");
    const volumes = Array.isArray(feature?.volumes) ? feature.volumes : [];
    if (!name) continue;
    for (const [key, value] of Object.entries(feature.environment || {})) env.push([String(key), String(value)]);
    if (!volumes.length) continue;
    allFeatures.add(name);

    const states = [];
    for (const volume of volumes) {
      const target = volume.path;
      if (typeof target !== "string" || !target) die(`invalid volume entry in feature ${name}: path is required`);
      const source = volume.source == null ? target : volume.source;
      if (typeof source !== "string" || !source) die(`invalid volume entry in feature ${name}: source is required`);
      const enabled = boolValue(volume.enabled, true, name, "enabled");
      states.push(enabled);
      rows.push({
        feature: name,
        enabled,
        source,
        target,
        file: boolValue(volume.file, false, name, "file"),
        readOnly: boolValue(volume.read_only, false, name, "read_only"),
        create: source === "~" || source.startsWith("~/") || !path.isAbsolute(source),
      });
    }
    options.set(name, states.some(Boolean) && !states.every(Boolean) ? "mixed" : (states.every(Boolean) ? "on" : "off"));
  }
  return { allFeatures, options, rows, env };
}

function printOptions(image, model) {
  if (!model.options.size) return console.log(`Config for ${image} exposes no volume toggles.`);
  console.log(`Config volume toggles for ${image}:`);
  for (const name of [...model.options.keys()].sort()) {
    console.log(`  --${name} / --no-${name} (default: ${model.options.get(name)}; features: ${name})`);
  }
}

function setOverride(arg, name, state, model, overrides) {
  if (model.options.has(name)) overrides.set(name, state);
  else if (model.allFeatures.has(name)) {
    die(`option ${state ? `--${name}` : `--no-${name}`} is known but not included in config (feature(s): ${name})`);
  } else {
    die(`unknown option: ${arg}`);
  }
}

function parseArgs(argv, model) {
  const overrides = new Map();
  let extraGroups = process.env.CONTAGENT_EXTRA_GROUP_GIDS || "";
  let show = false;
  let help = false;
  let command = [];
  for (let i = 0; i < argv.length; i++) {
    const arg = argv[i];
    if (arg === "--") { command = argv.slice(i + 1); break; }
    if (arg === "-h" || arg === "--help") { help = true; break; }
    if (arg === "--show-options") { show = true; continue; }
    if (arg === "-c" || arg === "--config") {
      if (i + 1 >= argv.length) die(`${arg} requires a value`);
      i += 1;
      continue;
    }
    if (arg.startsWith("--config=")) continue;
    if (arg === "--extra-groups" || arg.startsWith("--extra-groups=")) {
      if (arg === "--extra-groups" && i + 1 >= argv.length) die("--extra-groups requires a value");
      const value = arg === "--extra-groups" ? argv[++i] : arg.split("=", 2)[1];
      extraGroups = extraGroups ? `${extraGroups},${value}` : value;
      continue;
    }
    if (arg.startsWith("--no-")) {
      if (arg.includes("=")) die(`unknown option: ${arg}`);
      setOverride(arg, arg.slice(5), false, model, overrides);
      continue;
    }
    if (arg.startsWith("--")) {
      if (arg.includes("=")) die(`unknown option: ${arg}`);
      setOverride(arg, arg.slice(2), true, model, overrides);
      continue;
    }
    if (arg.startsWith("-")) die(`unknown option: ${arg}`);
    command = argv.slice(i);
    break;
  }
  return { command, show, help, extraGroups, overrides };
}

function mountPlan(rows, overrides, home, base) {
  const byTarget = new Map();
  for (const row of rows) {
    const enabled = overrides.has(row.feature) ? overrides.get(row.feature) : row.enabled;
    if (!enabled) continue;
    const source = resolvePath(row.source, home, base);
    const target = resolvePath(row.target, home, base);
    const candidates = byTarget.get(target) || [];
    const candidate = { source, file: row.file, readOnly: row.readOnly, create: row.create };
    if (!candidates.some((c) => JSON.stringify(c) === JSON.stringify(candidate))) candidates.push(candidate);
    byTarget.set(target, candidates);
  }

  const specs = [];
  const gids = [];
  for (const [target, candidates] of byTarget.entries()) {
    let chosen = candidates.find((c) => fs.existsSync(c.source));
    if (!chosen) {
      chosen = candidates.find((c) => c.create);
      if (!chosen) die(`no existing source found for target ${target} among ${candidates.length} candidates`);
      if (chosen.file) {
        fs.mkdirSync(path.dirname(chosen.source), { recursive: true });
        fs.closeSync(fs.openSync(chosen.source, "a"));
      } else {
        fs.mkdirSync(chosen.source, { recursive: true });
      }
    }
    try {
      const st = fs.statSync(chosen.source);
      const gid = String(st.gid);
      if (st.isSocket() && !gids.includes(gid)) gids.push(gid);
    } catch { /* ignore */ }
    specs.push(`${chosen.source}:${target}${chosen.readOnly ? ":ro" : ""}`);
  }
  return { specs, gids };
}

function extraGroupSpecs(csv) {
  const gids = [];
  for (const token of csv ? csv.split(",") : []) {
    const gid = token.trim();
    if (!gid) continue;
    if (!/^\d+$/.test(gid)) warn(`ignoring non-numeric extra group gid: ${gid}`);
    else if (!gids.includes(gid)) gids.push(gid);
  }
  return gids.map((gid) => `g${gid}:${gid}`).join(",");
}

function main() {
  if (cp.spawnSync("docker", ["--version"], { stdio: "ignore" }).error) die("docker is required");
  const image = process.env.CONTAGENT_IMAGE || "contagent:latest";
  const home = process.env.HOME || die("HOME must be set");
  const cwd = process.cwd();
  const { config, base } = loadConfig(configPath(process.argv.slice(2)), image);
  const model = buildModel(config);
  const parsed = parseArgs(process.argv.slice(2), model);
  if (parsed.help || parsed.show) {
    if (parsed.help) process.stdout.write(`${USAGE}\n`);
    printOptions(image, model);
    return;
  }

  const args = ["run", "--rm", "--workdir", cwd, "--volume", `${cwd}:${cwd}`];
  if (process.stdin.isTTY && process.stdout.isTTY) args.push("--interactive", "--tty");
  const env = [
    ["CONTAGENT_USERNAME", process.env.USER || id("-un")],
    ["CONTAGENT_GROUPNAME", id("-gn")],
    ["CONTAGENT_UID", id("-u")],
    ["CONTAGENT_GID", id("-g")],
    ["CONTAGENT_HOME", home],
    ...(process.env.TERM ? [["TERM", process.env.TERM]] : []),
    ...(process.env.COLORTERM ? [["COLORTERM", process.env.COLORTERM]] : []),
    ...model.env,
  ];
  for (const [key, value] of env) if (key) args.push("--env", `${key}=${value}`);
  const mounts = mountPlan(model.rows, parsed.overrides, home, base);
  for (const spec of mounts.specs) args.push("--volume", spec);
  const sshSock = process.env.SSH_AUTH_SOCK || "";
  if (sshSock && isSocket(sshSock)) args.push("--volume", `${sshSock}:${sshSock}`, "--env", `SSH_AUTH_SOCK=${sshSock}`);
  else warn("SSH agent not available; SSH auth forwarding disabled");
  const groupCsv = [parsed.extraGroups, ...mounts.gids].filter(Boolean).join(",");
  const groups = extraGroupSpecs(groupCsv);
  if (groups) args.push("--env", `CONTAGENT_EXTRA_GROUP_SPECS=${groups}`);

  const run = cp.spawnSync("docker", [...args, image, ...parsed.command], { stdio: "inherit" });
  if (run.error) die(run.error.message);
  process.exit(typeof run.status === "number" ? run.status : 1);
}

main();

File: contagent.py
#!/usr/bin/env python3

import json
import os
import re
import shutil
import stat
import subprocess
import sys
import tempfile
from pathlib import Path

USAGE = """Usage: ./contagent.py [options] [--] [command ...]

Options:
  -c, --config CONFIG         Runtime config path (default: .contagent.yaml)
  --<feature>                 Enable volume mounts for an image feature
  --no-<feature>              Disable volume mounts for an image feature
  --show-options              Show config-defined --<feature>/--no-<feature> toggles and exit
  --extra-groups <gid[,gid]>  Append supplementary group GIDs for this run
  -h, --help                  Show this help
"""

EMBEDDED_CONFIG = "/usr/local/share/contagent/contagent.yaml"


def die(msg: str) -> None:
    print(f"ERROR: {msg}", file=sys.stderr)
    raise SystemExit(1)


def warn(msg: str) -> None:
    print(f"WARN: {msg}", file=sys.stderr)


def is_socket(path: str) -> bool:
    try:
        return stat.S_ISSOCK(os.stat(path).st_mode)
    except OSError:
        return False


def resolve_path(raw: str, home: str, base: str) -> str:
    if raw == "~" or raw.startswith("~/"):
        return home + raw[1:]
    return raw if os.path.isabs(raw) else os.path.join(base, raw)


def bool_value(value: object, default: bool, feature: str, field: str) -> bool:
    if value is None:
        return default
    if not isinstance(value, bool):
        die(f"invalid volume entry in feature {feature}: {field} must be boolean")
    return value


def embedded_config(image: str) -> str:
    p = subprocess.run(
        ["docker", "run", "--rm", "--entrypoint", "cat", image, EMBEDDED_CONFIG],
        text=True,
        capture_output=True,
    )
    if p.returncode != 0:
        die(f"failed to extract default contagent config from {image}")
    return p.stdout


def yq_json(path: Path, image: str) -> dict:
    if shutil.which("yq"):
        attempts = [["yq", "-o=json", ".", str(path)], ["yq", ".", str(path)]]
        inputs = [None, None]
    else:
        warn(f"yq not found; falling back to yq from {image}")
        attempts = [
            ["docker", "run", "--rm", "-i", "--entrypoint", "yq", image, "-o=json", ".", "-"],
            ["docker", "run", "--rm", "-i", "--entrypoint", "yq", image, ".", "-"],
        ]
        inputs = [path.read_text(), path.read_text()]

    for cmd, data in zip(attempts, inputs):
        p = subprocess.run(cmd, input=data, text=True, capture_output=True)
        if p.returncode == 0:
            try:
                return json.loads(p.stdout)
            except Exception:
                pass
    die(f"failed to parse {path}")


def config_path(argv: list[str]) -> Path:
    config = ".contagent.yaml"
    i = 0
    while i < len(argv):
        arg = argv[i]
        if arg == "--":
            break
        if arg in ("-c", "--config"):
            if i + 1 >= len(argv):
                die(f"{arg} requires a value")
            config = argv[i + 1]
            i += 2
            continue
        if arg.startswith("--config="):
            config = arg.split("=", 1)[1]
        elif not arg.startswith("-"):
            break
        i += 1
    return Path(config)


def load_config(path: Path, image: str) -> tuple[dict, Path]:
    if not path.exists() or path.stat().st_size == 0:
        path.parent.mkdir(parents=True, exist_ok=True)
        data = embedded_config(image)
        if not data:
            die("embedded default contagent config is empty")
        path.write_text(data)

    config = yq_json(path, image)
    with tempfile.NamedTemporaryFile("w+", delete=False) as f:
        f.write(embedded_config(image))
        embedded_path = Path(f.name)
    try:
        embedded = yq_json(embedded_path, image)
    finally:
        embedded_path.unlink(missing_ok=True)

    config_id = config.get("image-hash")
    embedded_id = embedded.get("image-hash")
    if embedded_id and config_id != embedded_id:
        warn(f"{path} was not generated from {image}")
    return config, path.parent.resolve()


def build_model(config: dict) -> dict:
    options: dict[str, str] = {}
    rows: list[dict] = []
    env: list[tuple[str, str]] = []
    all_features: set[str] = set()

    for feature in config.get("features", []) if isinstance(config, dict) else []:
        name = str(feature.get("name") or "")
        volumes = feature.get("volumes") or []
        if not name:
            continue
        env.extend((str(k), str(v)) for k, v in (feature.get("environment") or {}).items())
        if not volumes:
            continue
        all_features.add(name)
        states: list[bool] = []

        for volume in volumes:
            target = volume.get("path")
            if not isinstance(target, str) or not target:
                die(f"invalid volume entry in feature {name}: path is required")
            source = volume.get("source") if volume.get("source") is not None else target
            if not isinstance(source, str) or not source:
                die(f"invalid volume entry in feature {name}: source is required")
            enabled = bool_value(volume.get("enabled"), True, name, "enabled")
            states.append(enabled)
            rows.append({
                "feature": name,
                "enabled": enabled,
                "source": source,
                "target": target,
                "file": bool_value(volume.get("file"), False, name, "file"),
                "read_only": bool_value(volume.get("read_only"), False, name, "read_only"),
                "create": source == "~" or source.startswith("~/") or not os.path.isabs(source),
            })
        options[name] = "mixed" if any(states) and not all(states) else ("on" if all(states) else "off")
    return {"all_features": all_features, "options": options, "rows": rows, "env": env}


def print_options(image: str, model: dict) -> None:
    if not model["options"]:
        print(f"Config for {image} exposes no volume toggles.")
        return
    print(f"Config volume toggles for {image}:")
    for name in sorted(model["options"]):
        print(f"  --{name} / --no-{name} (default: {model['options'][name]}; features: {name})")


def set_override(arg: str, name: str, state: bool, model: dict, overrides: dict) -> None:
    if name in model["options"]:
        overrides[name] = state
    elif name in model["all_features"]:
        flag = f"--{name}" if state else f"--no-{name}"
        die(f"option {flag} is known but not included in config (feature(s): {name})")
    else:
        die(f"unknown option: {arg}")


def parse_args(argv: list[str], model: dict) -> tuple[list[str], bool, bool, str, dict]:
    overrides: dict[str, bool] = {}
    extra_groups = os.environ.get("CONTAGENT_EXTRA_GROUP_GIDS", "")
    show_options = help_requested = False
    i = 0
    while i < len(argv):
        arg = argv[i]
        if arg == "--":
            return argv[i + 1:], show_options, help_requested, extra_groups, overrides
        if arg in ("-h", "--help"):
            return [], show_options, True, extra_groups, overrides
        if arg == "--show-options":
            show_options = True
        elif arg in ("-c", "--config"):
            if i + 1 >= len(argv):
                die(f"{arg} requires a value")
            i += 1
        elif arg.startswith("--config="):
            pass
        elif arg == "--extra-groups":
            if i + 1 >= len(argv):
                die("--extra-groups requires a value")
            extra_groups = f"{extra_groups},{argv[i + 1]}" if extra_groups else argv[i + 1]
            i += 1
        elif arg.startswith("--extra-groups="):
            value = arg.split("=", 1)[1]
            extra_groups = f"{extra_groups},{value}" if extra_groups else value
        elif arg.startswith("--no-"):
            if "=" in arg:
                die(f"unknown option: {arg}")
            set_override(arg, arg[5:], False, model, overrides)
        elif arg.startswith("--"):
            if "=" in arg:
                die(f"unknown option: {arg}")
            set_override(arg, arg[2:], True, model, overrides)
        elif arg.startswith("-"):
            die(f"unknown option: {arg}")
        else:
            return argv[i:], show_options, help_requested, extra_groups, overrides
        i += 1
    return [], show_options, help_requested, extra_groups, overrides


def mount_specs(rows: list[dict], overrides: dict, home: str, base: str) -> list[str]:
    by_target: dict[str, list[dict]] = {}
    order: list[str] = []
    for row in rows:
        enabled = overrides.get(row["feature"], row["enabled"])
        if not enabled:
            continue
        src = resolve_path(row["source"], home, base)
        dst = resolve_path(row["target"], home, base)
        candidate = {"source": src, **{k: row[k] for k in ("file", "read_only", "create")}}
        if dst not in by_target:
            by_target[dst] = []
            order.append(dst)
        if candidate not in by_target[dst]:
            by_target[dst].append(candidate)

    specs: list[str] = []
    for dst in order:
        candidates = by_target[dst]
        chosen = next((c for c in candidates if os.path.exists(c["source"])), None)
        if chosen is None:
            chosen = next((c for c in candidates if c["create"]), None)
            if chosen is None:
                die(f"no existing source found for target {dst} among {len(candidates)} candidates")
            if chosen["file"]:
                os.makedirs(os.path.dirname(chosen["source"]) or ".", exist_ok=True)
                Path(chosen["source"]).touch(exist_ok=True)
            else:
                os.makedirs(chosen["source"], exist_ok=True)
        specs.append(f"{chosen['source']}:{dst}" + (":ro" if chosen["read_only"] else ""))
    return specs


def extra_group_specs(csv: str) -> str:
    gids: list[str] = []
    for token in csv.split(",") if csv else []:
        gid = token.strip()
        if not gid:
            continue
        if not re.fullmatch(r"\d+", gid):
            warn(f"ignoring non-numeric extra group gid: {gid}")
        elif gid not in gids:
            gids.append(gid)
    return ",".join(f"g{gid}:{gid}" for gid in gids)


def main() -> None:
    if shutil.which("docker") is None:
        die("docker is required")
    image = os.environ.get("CONTAGENT_IMAGE", "contagent:latest")
    home = os.environ.get("HOME") or die("HOME must be set")
    cwd = os.getcwd()
    config, config_base = load_config(config_path(sys.argv[1:]), image)
    model = build_model(config)
    command, show, help_requested, groups, overrides = parse_args(sys.argv[1:], model)

    if help_requested or show:
        if help_requested:
            print(USAGE)
        print_options(image, model)
        return

    args = ["run", "--rm", "--workdir", cwd, "--volume", f"{cwd}:{cwd}"]
    if sys.stdin.isatty() and sys.stdout.isatty():
        args += ["--interactive", "--tty"]
    env = {
        "CONTAGENT_USERNAME": os.environ.get("USER") or subprocess.check_output(["id", "-un"], text=True).strip(),
        "CONTAGENT_GROUPNAME": subprocess.check_output(["id", "-gn"], text=True).strip(),
        "CONTAGENT_UID": subprocess.check_output(["id", "-u"], text=True).strip(),
        "CONTAGENT_GID": subprocess.check_output(["id", "-g"], text=True).strip(),
        "CONTAGENT_HOME": home,
    }
    for key in ("TERM", "COLORTERM"):
        if os.environ.get(key):
            env[key] = os.environ[key]
    env.update(dict(model["env"]))
    for key, value in env.items():
        args += ["--env", f"{key}={value}"]
    for spec in mount_specs(model["rows"], overrides, home, str(config_base)):
        args += ["--volume", spec]
    ssh_sock = os.environ.get("SSH_AUTH_SOCK", "")
    if ssh_sock and is_socket(ssh_sock):
        args += ["--volume", f"{ssh_sock}:{ssh_sock}", "--env", f"SSH_AUTH_SOCK={ssh_sock}"]
    else:
        warn("SSH agent not available; SSH auth forwarding disabled")
    group_specs = extra_group_specs(groups)
    if group_specs:
        args += ["--env", f"CONTAGENT_EXTRA_GROUP_SPECS={group_specs}"]
    os.execvp("docker", ["docker", *args, image, *command])


if __name__ == "__main__":
    main()
